### PR TITLE
[ros2] Add Pilz industrial motion planner as a solver option for move stages

### PR DESCRIPTION
**Pull request by sea-bass**
This PR adds a new solver that wraps the [Pilz industrial motion planner](https://github.com/ros-planning/moveit2/tree/main/moveit_planners/pilz_industrial_motion_planner). This allows using point-to-point (PTP), linear (LIN), and circular arc (CIRC) trajectories with move stages.

The biggest motivation for this was to define circular arcs for motion. You can see the tests for full details, but here is a stripped-down example:

```
pilz_solver = std::make_shared<solvers::Pilz>();
solver->setPlanner("CIRC");

auto cart_limits = pilz_industrial_motion_planner::CartesianLimit();
cart_limits.setMaxRotationalVelocity(1.0);
cart_limits.setMaxTranslationalVelocity(1.0);
cart_limits.setMaxTranslationalAcceleration(1.0);
cart_limits.setMaxTranslationalDeceleration(-1.0);
auto limits = pilz_industrial_motion_planner::LimitsContainer();
limits.setCartesianLimits(cart_limits);
solver->setLimits(limits);

auto stage = std::make_unique<stages::MoveTo>("move", pilz_solver);
stage->setGroup("panda_arm");
stage->setIKFrame("panda_hand");
stage->setGoal(goal_pose);
stage->setCircularArcConstraint({"center", center_pose})
```

Here's a quick video:

https://user-images.githubusercontent.com/4603398/199052890-1858ad93-ca2e-4de0-bcba-2440357f6ca8.mp4

Looking forward to feedback on this!

### CODE REVIEW

File: core/src/CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +43,7 @@ ament_target_dependencies(${PROJECT_NAME}
     moveit_ros_planning
     moveit_ros_planning_interface
     moveit_task_constructor_msgs
+	pilz_industrial_motion_planner
```
Comment (sea-bass) [inline core/src/CMakeLists.txt:46]: Fix tabs vs spaces

File: core/src/solvers/pilz.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	if (planner == "PTP") {
+		traj_gen_ =
+		    std::make_shared<pilz_industrial_motion_planner::TrajectoryGeneratorPTP>(robot_model, limits_, group_name);
+	} else if (planner == "LIN") {
+		traj_gen_ =
+		    std::make_shared<pilz_industrial_motion_planner::TrajectoryGeneratorLIN>(robot_model, limits_, group_name);
+	} else if (planner == "CIRC") {
+		traj_gen_ =
+		    std::make_shared<pilz_industrial_motion_planner::TrajectoryGeneratorCIRC>(robot_model, limits_, group_name);
+	}  // else case already handled in setPlanner() and empty checking logic
```
Comment (sea-bass) [inline core/src/solvers/pilz.cpp:128]: Is there a "cooler" C++ way of doing this?

File: core/src/stages/move_to.cpp [hunk 3]
```diff
@@ -238,7 +251,35 @@ bool MoveTo::compute(const InterfaceState& state, planning_scene::PlanningSceneP
...
+			path_constraints.name = arc_constraint_type;
+			moveit_msgs::msg::PositionConstraint pos_constraint;
+			pos_constraint.header.frame_id = arc_target_pose.header.frame_id;
+			pos_constraint.link_name = link->getName();
+			pos_constraint.constraint_region.primitive_poses.resize(1);
+			pos_constraint.constraint_region.primitive_poses[0] = arc_target_pose.pose;
+			pos_constraint.weight = 1.0;  // Shouldn't matter as it's the only constraint
+			path_constraints.position_constraints.resize(1);
+			path_constraints.position_constraints[0] = pos_constraint;
+		}
```
Comment (sea-bass) [inline core/src/stages/move_to.cpp:282]: This is the clunkiest part of the interface, but I couldn't come up with something better... Thoughts?

**Comment by sea-bass** [comment]
@JafarAbdi Do you know what may be happening here?

As the PR stands now, all the code builds fine, and the new tests run fine locally when setting up with the entire MoveIt2 setup. However, it seems we're failing to load the kinematics plugins for the Panda description on this CI.

File: core/src/stages/move_to.cpp [hunk 4]
```diff
@@ -67,6 +67,8 @@ MoveTo::MoveTo(const std::string& name, const solvers::PlannerInterfacePtr& plan
 	p.declare<std::string>("group", "name of planning group");
 	p.declare<geometry_msgs::msg::PoseStamped>("ik_frame", "frame to be moved towards goal pose");
 	p.declare<boost::any>("goal", "goal specification");
+	p.declare<std::pair<std::string, geometry_msgs::msg::PoseStamped>>("arc_constraint", "arc constraint specification");
+
```
Comment (rhaschke) [inline core/src/stages/move_to.cpp:71]: Same here. Consider as property of Pilz planner!

File: core/src/stages/move_to.cpp [hunk 5]
```diff
@@ -238,7 +251,35 @@ bool MoveTo::compute(const InterfaceState& state, planning_scene::PlanningSceneP
 		add_frame(ik_pose_world, "ik frame");
 
 		// transform target pose such that ik frame will reach there if link does
-		target = target * ik_pose_world.inverse() * scene->getCurrentState().getGlobalLinkTransform(link);
+		const auto ik_pose_link = ik_pose_world.inverse() * scene->getCurrentState().getGlobalLinkTransform(link);
+		target = target * ik_pose_link;
+
+		// Add circular arc path constraints, if enabled.
+		// Note that this replaces all other path constraints.
```
Comment (rhaschke) [inline core/src/stages/move_to.cpp:258]: Why not _adding_ the constraint?

**Review by rhaschke — COMMENTED** [review]
I'm not yet convinced that we need a specific `PlannerInterface` for this. The Pilz planner is a normal planning plugin, which can be chosen instead of OMPL for the `PipelinePlanner`. I agree that this PR introduces a slight simplification in specifying circular paths. However, if needed at all, I suggest deriving this planner from `PipelinePlanner` and just adding the syntactic sugar...

**Comment by sea-bass** [comment]
> I'm not yet convinced that we need a specific `PlannerInterface` for this. The Pilz planner is a normal planning plugin, which can be chosen instead of OMPL for the `PipelinePlanner`. I agree that this PR introduces a slight simplification in specifying circular paths. However, if needed at all, I suggest deriving this planner from `PipelinePlanner` and just adding the syntactic sugar...

... oh, wow. This is so much easier. As you suggested, these lines alone will do the same thing:

```c++
auto pilz_planner = std::make_shared<mtc::solvers::PipelinePlanner>(node_, "pilz_industrial_motion_planner");
pilz_planner->setPlannerId("CIRC");

auto stage = std::make_unique<mtc::stages::MoveTo>("move circle center", pilz_planner);
stage->setGroup("panda_arm");
stage->setIKFrame("panda_hand");
stage->setGoal(goal_pose);

moveit_msgs::msg::PositionConstraint pos_constraint;
pos_constraint.header.frame_id = center_pose.header.frame_id;
pos_constraint.link_name = hand_frame;
pos_constraint.constraint_region.primitive_poses.resize(1);
pos_constraint.constraint_region.primitive_poses[0] = center_pose.pose;
pos_constraint.weight = 1.0;
moveit_msgs::msg::Constraints path_constraints;
path_constraints.name = "center";
path_constraints.position_constraints.resize(1);
path_constraints.position_constraints[0] = pos_constraint;
stage->setPathConstraints(path_constraints);
```

Thank you for the feedback, and I think I'll go ahead and close this PR.

**Pull request closed without merging**